### ISSUE TITLE
Three-class likelihoods --> single binding probability?

### ISSUE THREAD

**Issue by rmovva** [davidaknowles/tf_net#4]
Hi again,

I was wondering how you were going from training on a labels matrix with three columns ((100,0,0), (0,1,0), or (0,0,1)) to the number p that's outputted from the training function. P is generally some negative number which you then apply a sigmoid transform to -- that step makes sense to get the probability itself, but I was wondering what's going on inside the train_func that's computing a single number from an array of size 3 with each of the probabilities for U, A, and B.

Here's the code snippet I'm referencing (in train.py):

```
# Run training
[p,neglike]=train_test_func( x_conv, x_flat, y_mat )
neglikes.append(float(neglike))

# Record labels and predictions for non-A for AUC calculations at end of epoch
for i in range(len(y)):
    if not y=="A":
        labs.append( 1 if y[i]=="B" else 0 )
        preds.append( expit(p[i]) )
```

Thanks!
Rajiv

**Comment by RishabGargeya** [comment]
I'm a bit confused by this as well.
Through my understanding, your TF_Net network has a last layer "b" that outputs a single value "logit_p" as the entire output of the network.

You then calculate the following values:
```
sml=T.nnet.sigmoid( - logit_p ) 
s1ml=T.nnet.sigmoid( 1.0 -logit_p ) 
p=T.stack( ( sml, s1ml - sml, 1.0 - s1ml ), axis=1)
```
where p stacks the above three values together.

I'm guessing sml -> U, (s1ml-sml) -> A, and (1.0 - s1ml) -> B?

But in the following lines:
```
# Record labels and predictions for non-A for AUC calculations at end of epoch
for i in range(len(y)):
    if not y=="A":
        labs.append( 1 if y[i]=="B" else 0 )
        preds.append( expit(p[i]) )
```

p[i] should retrieved the stack (array) of three values for that example __i__ in the minibatch, correct?
expit should only run the logistic function on these three values elementwise.
How do you translate these arrays of three values into a probability?

**Comment by davidaknowles** [comment]
This is problem with putting the code and not the math I guess! So g is the univariate output of the net and lives on the whole real line (positive and negative). Rishab's mapping to labels is correct. The likelihood is derived from P(U)=s(-g), P(A)=s(1-g)-s(-g), P(B)=1.0-s(1-g) where s is the logistic function. 

The prediction `expit(p[i])` is s(g)=1-P(U), i.e. the probability of being either A or B (you could take P(B) but since both s(g) and P(B) are monotonic increasing in g I don't think this would make any difference under the metrics used by the challenge). 

Does that clear things up?

**Comment by rmovva** [comment]
Thanks, David? But how do you get g? From my understanding it's the logistic function applied to p (variable called logit_p in your code), but what's p? Fundamentally I'm still confused on how you're going from an array of length 3 to a single output.

**Comment by davidaknowles** [comment]
g (logit_p in the code, should change that!) is what I'm calling the univariate output of the neural net. From g you calculate p according the to snippet Rishab included, and then the (negative) likelihood is calculated by the line 
```
neg_like=-( y * np.log(p + 1.0e-20) ).sum()
```
where y is the labels matrix you mention earlier.

**Comment by davidaknowles** [comment]
I've put a bit more detail on the underling math here
https://github.com/davidaknowles/tf_net/blob/master/METHODS.ipynb
so I'll close this.
